connectivity/mqtt/client_common.c

```diff
@@ -0,0 +1,519 @@
+        start = i + 1;
+      }
+    } else if (str[i] == '@') {
+      if (i == start) {
+        goto cleanup;
+      }
+      have_auth = true;
+      if (username_or_host) {
+        /* Must be of form socks5h://username:password@... */
+        username = username_or_host;
+        username_or_host = NULL;
+
+        len = i - start;
+        password = malloc(len + 1);
+        if (!password) {
+          fprintf(stderr, "Error: Out of memory.\n");
+          goto cleanup;
+        }
+        memcpy(password, &(str[start]), len);
+        password[len] = '\0';
+        start = i + 1;
+      } else {
+        /* Haven't seen a : yet, so must be of form
+         * socks5h://username@... */
+        if (username) {
+          /* Already got a username, must be malformed. */
+          goto cleanup;
+        }
+        len = i - start;
+        username = malloc(len + 1);
+        if (!username) {
+          fprintf(stderr, "Error: Out of memory.\n");
+          goto cleanup;
+        }
+        memcpy(username, &(str[start]), len);
+        username[len] = '\0';
+        start = i + 1;
+      }
+    }
+  }
+
+  /* Deal with remainder */
+  if (i > start) {
+    len = i - start;
+    if (host) {
+      /* Have already seen a @ , so this must be of form
+       * socks5h://username[:password]@host:port */
+      port = malloc(len + 1);
+      if (!port) {
+        fprintf(stderr, "Error: Out of memory.\n");
+        goto cleanup;
+      }
+      memcpy(port, &(str[start]), len);
+      port[len] = '\0';
+    } else if (username_or_host) {
+      /* Haven't seen a @ before, so must be of form
+       * socks5h://host:port */
+      host = username_or_host;
+      username_or_host = NULL;
+      port = malloc(len + 1);
+      if (!port) {
+        fprintf(stderr, "Error: Out of memory.\n");
+        goto cleanup;
+      }
+      memcpy(port, &(str[start]), len);
+      port[len] = '\0';
+    } else {
+      host = malloc(len + 1);
+      if (!host) {
+        fprintf(stderr, "Error: Out of memory.\n");
+        goto cleanup;
+      }
+      memcpy(host, &(str[start]), len);
+      host[len] = '\0';
+    }
+  }
+
+  if (!host) {
+    fprintf(stderr, "Error: Invalid proxy.\n");
+    goto cleanup;
+  }
+
+  if (mosquitto__urldecode(username)) {
+    goto cleanup;
+  }
+  if (mosquitto__urldecode(password)) {
+    goto cleanup;
+  }
+  if (port) {
+    port_int = atoi(port);
+    if (port_int < 1 || port_int > 65535) {
+      fprintf(stderr, "Error: Invalid proxy port %d\n", port_int);
+      goto cleanup;
+    }
+    free(port);
+  } else {
+    port_int = 1080;
+  }
+
+  cfg->socks_config->socks5_username = username;
+  cfg->socks_config->socks5_password = password;
+  cfg->socks_config->socks5_host = host;
+  cfg->socks_config->socks5_port = port_int;
+
+  return 0;
+cleanup:
+  if (username_or_host) free(username_or_host);
+  if (username) free(username);
+  if (password) free(password);
+  if (host) free(host);
+  if (port) free(port);
```

Comment:
Actually just `free()` without checking if it's NULL would be fine.
If it's NULL, no operation will be performed.